content/300-guides/300-upgrade-guides/200-upgrading-versions/030-upgrading-to-prisma-4.mdx

```diff
@@ -0,0 +1,431 @@
+1. You might now encounter a type error or runtime validation error when you use `Prisma.DbNull`, `Prisma.JsonNull`, or `Prisma.AnyNull` with MongoDB. This was never valid, but was silently accepted prior to Prisma 4. You need to review your data and change these fields to `null`.
+1. If you pass in dynamic JSON to a JSON column in Prisma Client (for example `prisma.findMany({where: { jsonColumn: someJson } })`), then you must check that `someJson`cannot be the string "DBNull", "JsonNull", or "AnyNull". If it is any of these values, then the query will return different results in version 4.0.0.
+
+### Default fields on composite types in MongoDB
+
+From version 4.0.0, if you carry out a database read on a composite type when all of the following conditions are true, then Prisma Client inserts the default value into the result.
+
+Conditions:
+
+- A field on the composite type is required, and
+- this field has a default value, and
+- this field is not present in the returned document or documents.
+
+This behavior is now consistent with the behavior for model fields.
+
+To learn more, see [Default values on required composite types](/concepts/components/prisma-client/composite-types#default-values-on-required-composite-types).
+
+#### Upgrade path
+
+If you currently rely on a return value of `null`, then you need to refactor your code to handle the default value that is now returned in Prisma 4.
+
+### Rounding errors on big numbers in SQLite
+
+SQLite is a loosely-typed database. If your schema has a field with type `Int`, then Prisma prevents you from inserting a value larger than an integer. However, nothing prevents the database from directly accepting a bigger number. These manually-inserted big numbers cause rounding errors when queried.
+
+To avoid this problem, Prisma version 4.0.0 and later checks numbers on the way out of the database to verify that they fit within the boundaries of an integer. If a number does not fit, then Prisma throws a P2023 error, such as:
+
+```
+Inconsistent column data: Conversion failed:
+Value 9223372036854775807 does not fit in an INT column,
+try migrating the 'int' column type to BIGINT
+```
+
+#### Upgrade path
+
+If you use Prisma in conjunction with SQLite, then you need to find any code that queries `Int` fields and ensure that it handles any P2023 errors that might be returned.
+
+### Prisma no longer exports `Prisma.dmmf.schema` into the generated Prisma Client
+
+From version 4.0.0, Prisma no longer exports `Prisma.dmmf.schema` into the generated Prisma Client. This makes the generated Prisma Client much more efficient, and also avoids some memory leaks with Jest.
+
+Note:
+
+- This change does not affect the DMMF that Prisma passes to the generators.
+- You can use `getDmmf()`from `@prisma/internals` to access the schema property.
+- We still export `Prisma.dmmf.datamodel` into the generated Prisma Client.
+
+### Index configuration
```

Comment:
This is _the_ big new feature that triggered Prisma 4. I suggest we move this up a bit in this list here as well.